### ISSUE TITLE
[Feature request] Ability to reimplement backprop.

### ISSUE THREAD

**Issue by alexmonroe87** [tensorflow/tensorflow#6422]
Currently there is no straightforward way to do things like implementing different learning rates for different layers. Much less apply learning rate masks for individual unit adjustment.

Torch allows you to easily reimplement backprop however you want. TF seems limited to this: https://www.tensorflow.org/api_docs/python/framework/core_graph_data_structures#Graph.gradient_override_map

**Comment by yaroslavvb** [comment]
What is the abstraction used in Torch/how would it look in TensorFlow?

**Comment by vrv** [comment]
@alexmonroe87 while there is probably a good feature request here, we need this to be a lot more precise for it to be useful.

One can use different instantiations of optimizers for variables with different learning rates to get per-layer learning rates.  I also don't know what a 'learning rate mask' is, so I'm not sure why it's hard.

It would be helpful to detail the use case more.  It's possible that what you want to do is possible in TF, but may require a little effort to think more declaratively, and we can't know unless we see a more concrete use case.

**Comment by alexmonroe87** [comment]
@vrv Using different optimizer instances for different variables would lead to the problem of computing the same gradients multiple times and probably some other issues. Might also not work as intended with an optimizer like Adam.

What I'm essentially asking for the is the ability to scale a gradient on a per neuron basis with a mask, or as in the use case of individual layers having their own learning rate, a scalar.

Simple example:
Some layer contains 5 neurons and you apply this mask: [1 1 2 1 1] to it, when doing backprop the middle neuron gets updated with twice the learning rate. Though in practice the learning rate will just be made 1 and the mask will specify it.

**Comment by vrv** [comment]
@alexmonroe87 an example where using different optimizer instances doesn't work would be useful -- I don't think we'd compute the same gradients multiple times if every variable is associated with one optimizer.  "Might not work as intended" is not sufficient evidence for a new feature request :).

As for your mask idea, you can do that today:  Let's say you have a single variable you want to optimize with a mask, using a GradientDescentOptimizer (for simplicity).  Here's my untested semi-TF code:

```
    v = tf.Variable(...)

    # Define your mask here, matching the shape of the variable.  E.g., could be [1.0, 1.0, 2.0, 1.0, 1.0]
    # if 'v' was a vector of 5.
    mask = tf.constant(..., dtype=tf.float32)

    # Base learning rate
    optimizer = tf.GradientDescentOptimizer(learning_rate)

    # Compute the gradients and have them returned
    grads_and_vars = optimizer.compute_gradients(loss, var_list=[v], ...)

    # Scale the gradient by the mask you want
    masked_grad_and_vars = [(tf.multiply(mask, g), v) for g, v in grads_and_vars]
    
    # Apply the scaled gradient
    optimizer.apply_gradients(masked_grad_and_vars)
```

The only difference between this and normal code is the addition of the 'mask' and the 'masked_grad_and_vars'.  Everything else is the same.

Can you explain why this wouldn't work for you?

**Comment by alexmonroe87** [comment]
@vrv Scaling incoming gradients doesn't work for optimizers that adaptively rescale/normalize the gradients. Such as Adam.

**Comment by yaroslavvb** [comment]
There's a solution [here](http://stackoverflow.com/a/39793644/419116) that works for adam (specify different optimizer for each layer)

**Comment by vrv** [comment]
@alexmonroe87 reimplementing backprop for a given op would only allow you to control the output prior to the optimizer, so I still don't understand your feature request. 

I'm going to close this particular issue since the request is still unclear after repeated requests for clarfiication.

If you can provide *concrete* examples of what is not possible / what is hard, (or more evidence of why something is impossible), we can reopen.
